feat(alert): don't timeout if isMouseOver

redallen opened the pull request:
<!-- What changes are being made? Please link the issue being addressed. -->
**What**: Wait until the Alert is timed out AND the mouse is not over it AND the keyboard isn't focusing an element inside of it. After the mouse exits wait at least `timeoutAnimation` seconds. Closes #5489

<!-- Are there any upstream issues or separate issues you need to reference? -->
**Additional issues**:

patternfly-build commented:
PF4 preview: https://patternfly-react-pr-5566.surge.sh

christianvogt commented on packages/react-core/src/components/Alert/Alert.tsx:
```diff
@@ -91,12 +94,12 @@ export const Alert: React.FunctionComponent<AlertProps> = ({
+  React.useEffect(() => {
+    timeout = timeout === true ? 8000 : Number(timeout);
+    if (timeout > 0) {
+      setTimeout(() => setTimedOut(true), timeout);
```
You need to clear the timeout on unmount otherwise you'll end up updating state on an unmounted component if the component is unmounted before the timeout completes.

christianvogt commented on packages/react-core/src/components/Alert/Alert.tsx:
```diff
@@ -107,49 +110,54 @@ export const Alert: React.FunctionComponent<AlertProps> = ({
+  };
+
+  if (timedOut && !isMouseOver) {
+    onTimeout();
```
Do not execute side effects in `render`. Use an effect.

christianvogt reviewed (commented):
I'd prefer to see an approach where if the user interacts with the alert, the current timeout is first cleared. When the user mouses away, a timeout resumes but at a much lower time. Perhaps `Math.max(timeout / 2, 3)` or even just a small fixed time. The reason for this is to not immediately hide the alert due to a mouse slip.

Also, we shouldn't hide the alert if the user focuses on an action via keyboard. You'll need to track focus in addition to mouse events.

redallen commented:
I discussed this with @mcarrano and we decided on a small 3s timeout to prevent mouseslips. I addressed all your feedback, the keyboard focus was rather tricky.

christianvogt commented on packages/react-core/src/components/Alert/Alert.tsx:
```diff
@@ -91,12 +100,32 @@ export const Alert: React.FunctionComponent<AlertProps> = ({
+    return () => document.removeEventListener('focus', onDocumentFocus, true);
+  }, [containsFocus]);
+  const dismissed = timedOut && timedOutAnimation && !isMouseOver && !containsFocus;
+  React.useEffect(onTimeout, [dismissed]);
```
You should only call `onTimeout` when `dismissed` is `true` otherwise you notify on first render.
```suggestion
  React.useEffect(() => dismissed && onTimeout(), [dismissed]);
```

christianvogt commented on packages/react-core/src/components/Alert/Alert.tsx:
```diff
@@ -107,49 +136,57 @@ export const Alert: React.FunctionComponent<AlertProps> = ({
+    onMouseLeave(ev);
+  };
+
+  if (dismissed) {
+    onTimeout();
```
You can remove this now that it's done in an effect.

christianvogt commented on packages/react-core/src/components/Alert/Alert.tsx:
```diff
@@ -107,49 +136,57 @@ export const Alert: React.FunctionComponent<AlertProps> = ({
-  } else {
+  const myOnMouseLeave = (ev: React.MouseEvent<HTMLDivElement>) => {
+    setIsMouseOver(false);
+    setTimeout(() => setTimedOutAnimation(true), timeoutAnimation);
```
This timer needs to be cleared on mouse enter and restarted on mouse leave.
You also need to cancel the timer on component unmount.

redallen replied:
> restarted on mouse leave
Not sure how to do this part. I don't think it needs to be restarted on mouse leave since `const dismissed = timedOut && timedOutAnimation && !isMouseOver && !containsFocus;`

christianvogt replied:
looks like you changed the implementation here, so this won't be needed now

christianvogt commented on packages/react-core/src/components/Alert/Alert.tsx:
```diff
@@ -91,12 +100,32 @@ export const Alert: React.FunctionComponent<AlertProps> = ({
+          setTimedOutAnimation(false);
+        } else if (containsFocus) {
+          setContainsFocus(false);
+          setTimeout(() => setTimedOutAnimation(true), timeoutAnimation);
```
Cancel timer on unmount.

christianvogt commented on packages/react-core/src/components/Alert/Alert.tsx:
```diff
@@ -91,12 +96,57 @@ export const Alert: React.FunctionComponent<AlertProps> = ({
+
+  const [timedOut, setTimedOut] = useState(false);
+  const [timedOutAnimation, setTimedOutAnimation] = useState(true);
+  const [isMouseOver, setIsMouseOver] = useState<boolean | undefined>();
+  const [containsFocus, setContainsFocus] = useState<boolean | undefined>();
```
```suggestion
  const [isMouseOver, setIsMouseOver] = useState<boolean>(false);
  const [containsFocus, setContainsFocus] = useState<boolean>(false);
```

redallen replied:
This won't work because then this `useEffect`'s condition will needlessly happen on first render:
```
  React.useEffect(() => {
    if (containsFocus === false || isMouseOver === false) {
      const timer = setTimeout(() => setTimedOutAnimation(true), timeoutAnimation);
      return () => clearTimeout(timer);
    }
  }, [containsFocus, isMouseOver]);
```

christianvogt replied:
tricky tri-state booleans!

christianvogt commented:
tested it out and it seems to work as expected now

mcarrano reviewed (approved):
Looks good to me! Thanks @redallen .

tlabaj reviewed (approved):
lgtm
